### PR TITLE
output: allow records size 0 on flb event type logs processor

### PR DESCRIPTION
**Pull request by niedbalski**
### Description.

For full context, refer to #9553.

A chunk with no records is a valid scenario for logs (e.g., when all are dropped) and should either be added to the flush list or released immediately. However, the chunk should not attempt to flush.

This patch allows for the record size to be zero for `FLB_EVENT_TYPE_LOGS`.

I created a basic processor that calls `flb_mp_chunk_cobj_record_destroy` on each record, resulting in a chunk with no records (size = 0). When this happens, coroutine tasks do not run, and `fluentbit_storage_mem_chunks` continues to grow, causing a memory leak (refer to the issue report).

```c
#include <fluent-bit/flb_processor_plugin.h>

static int cb_init(struct flb_processor_instance *ins,
                   void *source_plugin_instance,
                   int source_plugin_type,
                   struct flb_config *config)
{
    return FLB_PROCESSOR_SUCCESS;
}

static int cb_exit(struct flb_processor_instance *ins, void *data)
{
    return FLB_PROCESSOR_SUCCESS;
}

static int cb_process_logs(struct flb_processor_instance *ins,
                           void *chunk_data, const char *tag, int tag_len){
    struct flb_mp_chunk_cobj *chunk_cobj = (struct flb_mp_chunk_cobj *) chunk_data;
    struct flb_mp_chunk_record *record;
    while (flb_mp_chunk_cobj_record_next(chunk_cobj, &record) == FLB_MP_CHUNK_RECORD_OK) {
        flb_mp_chunk_cobj_record_destroy(chunk_cobj, record);  
    }
    return FLB_PROCESSOR_SUCCESS;
}


static struct flb_config_map config_map[] = {
    {0}
};

struct flb_processor_plugin processor_template_plugin = {
    .name               = "chunk_leaks",
    .description        = "chunk_leaks processor",
    .cb_init            = cb_init,
    .cb_process_logs    = cb_process_logs,
    .cb_process_metrics = NULL,
    .cb_process_traces  = NULL,
    .cb_exit            = cb_exit,
    .config_map         = NULL,
    .flags              = 0
};

```

With this configuration 

```
service:
  http_server: on
pipeline:
  inputs:
    - name: dummy
      rate: 1000
      dummy: '{"message": "a message", "event_cluster": "raptor"}'
  outputs:
    - name: stdout
      match: '*'
      processors:
        logs:
            - name: chunk_leaks
```

Without this patch heap allocation graphs show continuous growth in heap memory allocations not being returned, with the chunk counter also incrementing in metrics. 

image

With this patch enabled, chunks are released 

image

**Testing**

- [X] Example configuration file for the change
- [X] Debug log output from testing the change
- [X] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Backporting**

- [X] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: include/fluent-bit/flb_output.h [hunk 1]
```diff
@@ -676,21 +676,13 @@ struct flb_output_flush *flb_output_flush_create(struct flb_task *task,
...
-            records = flb_mp_count(p_buf, p_size);
-            if (records == 0) {
-                flb_coro_destroy(coro);
-                flb_free(out_flush);
-                flb_free(p_buf);
-                return NULL;
-            }
-
-            tmp = flb_event_chunk_create(evc->type, records, evc->tag, flb_sds_len(evc->tag), p_buf, p_size);
+            tmp = flb_event_chunk_create(evc->type, flb_mp_count(p_buf, p_size), evc->tag, flb_sds_len(evc->tag), p_buf, p_size);
```
Comment (edsiper) [inline include/fluent-bit/flb_output.h:685]: flb_mp_count() can fail if there is a corrupted record, it's a bit safer to validate its return value out of the function. For the majority of use cases except for sds_len() we do not use function returns as arguments

**Review by edsiper — CHANGES_REQUESTED** [review]

File: include/fluent-bit/flb_output.h (hunk 1 above)
Comment (niedbalski) [reply include/fluent-bit/flb_output.h:685]: @edsiper sure, I can fix.

**Comment by niedbalski** [comment]
@edsiper / @cosmo0920 seems stable / no chunk leaks for 15m or so. Appreciated suggestions on testing if required.
image

**Review by cosmo0920 — APPROVED** [review]
Looks good and I have never imagined and though as not returned chunks cause memory leaks. This could be interesting behavior and we need to return when the size of chunks are zero. It won't be caused without processors but now we have processors. So, we need to change this behavior.

**Pull request merged**